pkg/middleware/instrumentation.go

```diff
@@ -0,0 +1,109 @@
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/justinas/alice"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var metricsHandler = NewInstrumentMetricHandlerWithDefaultRegistry()
+
+func NewInstrumentMetricHandlerWithDefaultRegistry() http.Handler {
+	return NewInstrumentMetricHandler(prometheus.DefaultRegisterer, prometheus.DefaultGatherer)
+}
+
+func NewInstrumentMetricHandler(registerer prometheus.Registerer, gatherer prometheus.Gatherer) http.Handler {
+	return promhttp.InstrumentMetricHandler(
+		registerer, promhttp.HandlerFor(gatherer, promhttp.HandlerOpts{}),
+	)
+}
+
+func NewRequestInstrumentorWithDefaultRegistry() alice.Constructor {
+	return NewRequestInstrumentor(prometheus.DefaultRegisterer)
+}
+
+func NewRequestInstrumentor(registerer prometheus.Registerer) alice.Constructor {
+	return func(next http.Handler) http.Handler {
+		// Counter for all requests
+		// This is bucketed based on the response code we set
+		counterHandler := func(next http.Handler) http.Handler {
+			return promhttp.InstrumentHandlerCounter(registerRequestsCounter(registerer), next)
+		}
+
+		// Gauge to all requests currently being handled
+		inFlightHandler := func(next http.Handler) http.Handler {
+			return promhttp.InstrumentHandlerInFlight(registerInflightRequestsGauge(registerer), next)
+		}
+
+		// The latency of all requests bucketed by HTTP method
+		durationHandler := func(next http.Handler) http.Handler {
+			return promhttp.InstrumentHandlerDuration(registerRequestsLatencyHistogram(registerer), next)
+		}
+
+		return alice.New(counterHandler, inFlightHandler, durationHandler).Then(next)
```

Comment:
These definitely seem like the most important toward getting good RED metrics 👍 

`InstrumentHandlerResponseSize` would be another one that might be worth us instrumenting in the future. (the RequestSize variant too maybe?)